File: src/c#/基础语法/ConsoleApp.BestPractices/AboutFile.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace ConsoleApp.BestPractices
{
    /// <summary>
    /// 一般用静态方法就行，如果对性能要求很高，才考虑打开后操作，操作完再关闭
    /// </summary>
    public class AboutFile
    {
        public static void RunDemo()
        {
            Console.WriteLine("控制台输入输出，随便输入个，回车结束:");
            var input = Console.ReadLine();
            Console.WriteLine($"你输入了：{input}");



            var filepath = "C:\\FileTest\\haha\\1.txt";
            var context = "内容";
            var lstContext = new List<string> { "A", "B" };
            var byteContext = Encoding.UTF8.GetBytes(context);

            //获取当前工作目录
            var workingDirectory = Environment.CurrentDirectory;
            
            //改变当前工作目录，说改就改，之后的文件操作的BasePath就是这个了，Java11没有这样的功能。
            Environment.CurrentDirectory = "c:\\windows";


            //仅当前目录
            var files1 = Directory.GetFiles("C:\\Windows\\", "*.exe");

            //包括子目录，可以指定多种筛选规则
            var files2 = Directory.GetFiles("C:\\Windows\\", "*.EXE", new EnumerationOptions
            {
                RecurseSubdirectories = true,
                MatchCasing = MatchCasing.PlatformDefault,
                AttributesToSkip = FileAttributes.Normal
            });

            Directory.CreateDirectory("C:\\FileTest\\haha"); //存在不报错

            if (!File.Exists(filepath))
            {
                //xx
            }


            File.AppendAllText(filepath, context);

            File.WriteAllText(filepath, context);
            File.WriteAllLines(filepath, lstContext);
            File.WriteAllBytes(filepath, byteContext);

            //文件在用时无权删除，如果要删除则需要包装下kernel32.dll

            //http://www.jeremyshanks.com/fastest-way-to-write-text-files-to-disk-in-c/
            using (var writer = new StreamWriter("C:\\bigtest.txt", true, Encoding.UTF8, 65536))
            {
                var temp = "12345".Repeat(1024);
                for (int i = 0; i < 1000000; i++)
                {
                    writer.WriteLine(temp);
                }
            }
            File.Delete("C:\\bigtest.txt");


            var cnt1 = File.ReadAllText(filepath);
            var cnt2 = File.ReadAllLines(filepath);
            var cnt3 = File.ReadAllBytes(filepath);


            File.Copy(filepath, filepath + ".txt", true);
            File.Delete(filepath);


            var tempPath = Path.Combine("C:", "a", "b", "c.txt"); //C:\a\b\c.txt 
            var filename = Path.GetFileName(tempPath);
            var fullpath = Path.GetFullPath(tempPath);
            var extend = Path.GetExtension(tempPath);
        }
    }
}

File: src/java/基础语法/src/com/demo/AboutFile.java
package com.demo;


import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;
import java.util.List;
import java.util.Scanner;

//io与nio的不同要写下，因为同时存在。我现在看到的是：nio已经可以很方便的满足我的要求了，没必要研究旧io了
public class AboutFile {
    public static void main(String[] args) throws Exception {
        System.out.println("控制台输入输出，随便输入个，回车结束:");
        var input = new Scanner(System.in).nextLine();
        System.out.println("你输入了：" + input);


        var filepath = "C:\\FileTest\\haha\\1.txt";
        var context = "内容";
        var lstContext = List.of("A", "B");
        var byteContext = context.getBytes(StandardCharsets.UTF_8);

        //获取当前工作目录
        var workingDirectory = System.getProperty("user.dir");

        //虽然能能改变这个配置，但Path.of等在启动时已经将其固化了，再改也不会用他。而C#可以很容的修改
        System.setProperty("user.dir", "C:\\windows");

        //仅当前目录
        var files1 = Files.list(Path.of("C:\\Windows\\")).filter(p -> p.toString().toLowerCase().endsWith(".exe"));

        //包括子目录，可以指定层次
        var files2 = Files.walk(Path.of("C:\\Windows\\"), 10);

        //包括子目录，可以指定层次，可以指定filter。path.endsWith与path.toString().endsWith不同
        var files3 = Files.find(Path.of("C:\\Windows\\"), 10,
                (filePath, fileAttr) -> filePath.toString().endsWith(".exe") && fileAttr.isRegularFile());

        // createDirectories 如果目录已经存在不会报错
        Files.createDirectories(Path.of("C:\\FileTest\\haha"));


        if (!Files.exists(Path.of(filepath))) {
            //xx
        }

        var path = Path.of(filepath);
        System.out.println(path.getParent()); // C:\FileTest\haha
        System.out.println(path.toFile().getParent()); // C:\FileTest\haha

        Files.write(Path.of(filepath), context.getBytes(StandardCharsets.UTF_8),
                //不存在则创建，存在则附加。
                StandardOpenOption.CREATE, StandardOpenOption.APPEND);

        Files.write(Path.of(filepath), context.getBytes(StandardCharsets.UTF_8));
        Files.write(Path.of(filepath), lstContext);

        //快速写大量内容
        try (var writer = Files.newBufferedWriter(Path.of("C:\\bigtest.txt"),
                StandardCharsets.UTF_8, StandardOpenOption.CREATE, StandardOpenOption.APPEND)) {
            var temp = "12345".repeat(1024);
            for (int i = 0; i < 100000; i++) {
                writer.append(temp);
                writer.newLine();
            }
        }
        Files.delete(Path.of("C:\\bigtest.txt"));

        //读取utf-8 with bom文件后，字符串开头自动加了个不可见的字符。
        //如有需要请自行去掉原文件的bom或者读取后替换下：tmp.replace("\uFEFF", "")，JDK不解决此问题
        var cnt1 = Files.readString(Path.of(filepath));
        var cnt2 = Files.readAllLines(Path.of(filepath));
        var cnt3 = Files.readAllBytes(Path.of(filepath));

        Files.copy(Path.of(filepath), Path.of(filepath + ".txt"), StandardCopyOption.REPLACE_EXISTING);
        Files.delete(Path.of(filepath));


        var tempPath = Path.of("C:", "a", "b", "c.txt"); //C:\a\b\c.txt
        var filename = tempPath.getFileName();
        var fullpath = tempPath.toAbsolutePath();

        //获取扩展名没有像C#里的Path.GetExtension(tempPath);
        // Guava：Files.getFileExtension(filepath);
        // Apache Commons IO：FilenameUtils.getExtension(filepath);
        int dotIndex = filepath.lastIndexOf('.');
        var extend = (dotIndex == -1) ? "" : filepath.substring(dotIndex + 1);

    }
}
